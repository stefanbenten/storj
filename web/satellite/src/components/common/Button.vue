// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <!-- if isDisabled check onPress in parent element -->
    <div
        v-bind:class="containerClassName"
        :style="style"
        v-on:click="onPress">
            <h1 v-bind:class="[isWhite ? 'label white' : 'label']">{{label}}</h1>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

// Custom button component with label
@Component(
    {
        props: {
            label: {
                type: String,
                default: 'Default'
            },
            width: {
                type: String,
                default: 'inherit'
            },
            height: {
                type: String,
                default: 'inherit'
            },
            isWhite: {
                type: Boolean,
                default: false
            },
            isDeletion: {
                type: Boolean,
                default: false
            },
            isDisabled: {
                type: Boolean,
                default: false
            },
            onPress: {
                type: Function,
                default: () => {
                    console.error('onPress is not reinitialized');
                }
            }
        },
        computed: {
            style: function () {
                return {width: this.$props.width, height: this.$props.height};
            },
            containerClassName: function () {
                if (this.$props.isDisabled) return 'container disabled';

                if (this.$props.isWhite) return 'container white';

                if (this.$props.isDeletion) return 'container red';
                
                return 'container';
            },
        }
    }
)

export default class Button extends Vue {
}
</script>

<style scoped lang="scss">
    .container {
        display: flex;
        align-items: center;
        justify-content: center;
        background-color: #2683FF;
        border-radius: 6px;
        cursor: pointer;

        &:hover {
            box-shadow: 0px 4px 20px rgba(35, 121, 236, 0.4);

            &.white {
                box-shadow: none;
                background-color: #2683FF;
                border: 1px solid #2683FF;

                .label {
                    color: white;
                }
            }

            &.red {
                box-shadow: none;
                background-color: transparent;

                .label {
                    color: #EB5757;
                }
            }

            &.disabled {
                box-shadow: none;
            }
        }

        .label {
            font-family: 'montserrat_medium';
			font-size: 16px;
			line-height: 23px;
            color: #fff;
        }
    }
    .container.white,
    .container.red {
        background-color: transparent;
        border: 1px solid #AFB7C1;

        .label {
            color: #354049;
        }
    }
    .container.disabled {
        background-color: #DADDE5;
        .label.white {
            color: #ACB0BC;
        }
    }
</style>
