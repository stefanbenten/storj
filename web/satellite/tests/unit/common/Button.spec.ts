// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import { mount, shallowMount } from '@vue/test-utils';
import Button from '@/components/common/Button.vue';
import * as sinon from 'sinon';

describe('Button.vue', () => {

	it('renders correctly', () => {

		const wrapper = shallowMount(Button);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with isWhite prop', () => {

		const wrapper = shallowMount(Button, {
			propsData: {
				isWhite: true
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with isDisabled prop', () => {

		const wrapper = shallowMount(Button, {
			propsData: {
				isDisabled: true
			}
		});

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with size and label props', () => {
		let label = 'testLabel';
		let width = '30px';
		let height = '20px';

		const wrapper = shallowMount(Button, {
			propsData: {label, width, height},
		});

		expect(wrapper.element.style.width).toMatch(width);
		expect(wrapper.element.style.height).toMatch(height);
		expect(wrapper.text()).toMatch(label);
	});

	it('renders correctly with default props', () => {

		const wrapper = shallowMount(Button);

		expect(wrapper.element.style.width).toMatch('inherit');
		expect(wrapper.element.style.height).toMatch('inherit');
		expect(wrapper.text()).toMatch('Default');
	});

	it('trigger onPress correctly', () => {
		let onPressSpy = sinon.spy();

		const wrapper = mount(Button, {
			propsData: {
				onPress: onPressSpy,
				isDisabled: false
			}
		});

		wrapper.find('div.container').trigger('click');

		expect(onPressSpy.callCount).toBe(1);
	});
});
