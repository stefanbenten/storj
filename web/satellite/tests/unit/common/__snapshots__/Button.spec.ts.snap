// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue renders correctly 1`] = `
<div class="container" style="width: inherit; height: inherit;">
  <h1 class="label">Default</h1>
</div>
`;

exports[`Button.vue renders correctly with isDisabled prop 1`] = `
<div class="container disabled" style="width: inherit; height: inherit;">
  <h1 class="label">Default</h1>
</div>
`;

exports[`Button.vue renders correctly with isWhite prop 1`] = `
<div class="container white" style="width: inherit; height: inherit;">
  <h1 class="label white">Default</h1>
</div>
`;
